[openssl] Build error on x64-windows

Copy issue body from D:/vcpkg/vcpkg/installed/vcpkg/issue_body.md

error: building openssl:x64-windows failed with: BUILD_FAILED
Elapsed time to handle openssl:x64-windows: 4.1 min
Please ensure you're using the latest port files with `git pull` and `vcpkg update`.
Then check for known issues at:
  https://github.com/microsoft/vcpkg/issues?q=is%3Aissue+is%3Aopen+in%3Atitle+openssl
You can submit a new issue at:
  https://github.com/microsoft/vcpkg/issues/new?title=[openssl]+Build+error+on+x64-windows&body=Copy+issue+body+from+D%3A%2Fvcpkg%2Fvcpkg%2Finstalled%2Fvcpkg%2Fissue_body.md

Package: openssl:x64-windows@3.2.0#1

**Host Environment**

- Host: x64-windows
- Compiler: MSVC 19.29.30146.0
-    vcpkg-tool version: 2023-12-12-1c9ec1978a6b0c2b39c9e9554a96e3e275f7556e
    vcpkg-scripts version: 839722725 2023-12-29 (2 days ago)

**To Reproduce**

`vcpkg install tesseract:x64-windows`
**Failure logs**

```
-- Downloading https://github.com/openssl/openssl/archive/openssl-3.2.0.tar.gz -> openssl-openssl-openssl-3.2.0.tar.gz...
-- Extracting source D:/vcpkg/vcpkg/downloads/openssl-openssl-openssl-3.2.0.tar.gz
-- Applying patch asm-comments.patch
-- Applying patch declspec-align.patch
-- Applying patch qt-msvc.patch
-- Applying patch script-prefix.patch
-- Applying patch windows/install-layout.patch
-- Applying patch windows/install-pdbs.patch
-- Applying patch unix/android-cc.patch
-- Applying patch unix/move-openssldir.patch
-- Applying patch unix/no-empty-dirs.patch
-- Applying patch unix/no-static-libs-for-shared.patch
-- Using source at D:/vcpkg/vcpkg/buildtrees/openssl/src/nssl-3.2.0-06d693f20a.clean
-- Downloading https://github.com/StrawberryPerl/Perl-Dist-Strawberry/releases/download/SP_5380_5361/strawberry-perl-5.38.0.1-64bit-portable.zip -> strawberry-perl-5.38.0.1-64bit-portable.zip...
-- Found external ninja('1.10.2').
-- Getting CMake variables for x64-windows
-- Downloading https://www.nasm.us/pub/nasm/releasebuilds/2.16.01/win64/nasm-2.16.01-win64.zip;https://gstreamer.freedesktop.org/src/mirror/nasm-2.16.01-win64.zip -> nasm-2.16.01-win64.zip...
-- Getting CMake variables for x64-windows
-- Downloading https://download.qt.io/official_releases/jom/jom_1_1_4.zip;https://mirrors.ocf.berkeley.edu/qt/official_releases/jom/jom_1_1_4.zip;https://mirrors.ukfast.co.uk/sites/qt.io/official_releases/jom/jom_1_1_4.zip -> jom_1_1_4.zip...
-- Prerunning x64-windows-dbg
-- Building x64-windows-dbg
-- Restarting build without parallelism
CMake Error at scripts/cmake/vcpkg_execute_build_process.cmake:134 (message):
    Command failed: "D:/VS Studio/VC/Tools/MSVC/14.29.30133/bin/Hostx64/x64/nmake.exe" /NOLOGO /G /U /F makefile install_dev install_modules INSTALL_PDBS=ON
    Working Directory: D:/vcpkg/vcpkg/buildtrees/openssl/x64-windows-dbg
    See logs for more information:
      D:\vcpkg\vcpkg\buildtrees\openssl\install-x64-windows-dbg-nmake-out.log
      D:\vcpkg\vcpkg\buildtrees\openssl\install-x64-windows-dbg-nmake-err.log

Call Stack (most recent call first):
  scripts/cmake/vcpkg_build_nmake.cmake:156 (vcpkg_execute_build_process)
  ports/openssl/windows/portfile.cmake:69 (vcpkg_build_nmake)
  ports/openssl/portfile.cmake:77 (include)
  scripts/ports.cmake:170 (include)



```
<details><summary>D:\vcpkg\vcpkg\buildtrees\openssl\install-x64-windows-dbg-nmake-err.log</summary>

```
cl : Command line warning D9025 : overriding '/Zi' with '/Z7'
NMAKE : fatal error U1077: '"D:\VS Studio\VC\Tools\MSVC\14.29.30133\bin\Hostx64\x64\cl.exe"' : return code '0x2'
Stop.
NMAKE : fatal error U1077: '"D:\VS Studio\VC\Tools\MSVC\14.29.30133\bin\Hostx64\x64\nmake.exe"' : return code '0x2'
Stop.
```
</details>

<details><summary>D:\vcpkg\vcpkg\buildtrees\openssl\install-x64-windows-dbg-nmake-out.log</summary>

```
	"D:\VS Studio\VC\Tools\MSVC\14.29.30133\bin\Hostx64\x64\nmake.exe" /LU                 depend
	"D:\VS Studio\VC\Tools\MSVC\14.29.30133\bin\Hostx64\x64\nmake.exe" /LU                 _build_libs
	"D:\VS Studio\VC\Tools\MSVC\14.29.30133\bin\Hostx64\x64\cl.exe"  /Zi /Fdossl_static.pdb /Gs0 /GF /Gy /MDd -nologo -DWIN32 -D_WINDOWS -W3 -utf-8 -MP  -D_DEBUG -MDd -Z7 -Ob0 -Od -RTC1 -I"crypto" -I"." -I"include" -I"providers\common\include" -I"providers\implementations\include" -D"L_ENDIAN" -D"OPENSSL_PIC" -D"OPENSSLDIR=\"D:\\vcpkg\\vcpkg\\packages\\openssl_x64-windows\\debug\"" -D"ENGINESDIR=\"D:\\vcpkg\\vcpkg\\packages\\openssl_x64-windows\\debug\\lib\\engines-3\"" -D"MODULESDIR=\"D:\\vcpkg\\vcpkg\\packages\\openssl_x64-windows\\debug\\bin\"" -D"OPENSSL_BUILDING_OPENSSL" -D"OPENSSL_SYS_WIN32" -D"WIN32_LEAN_AND_MEAN" -D"UNICODE" -D"_UNICODE" -D"_CRT_SECURE_NO_DEPRECATE" -D"_WINSOCK_DEPRECATED_NO_WARNINGS" -D"DEBUG" -D"_DEBUG"   -DAES_ASM -DBSAES_ASM -DCMLL_ASM -DECP_NISTZ256_ASM -DGHASH_ASM -DKECCAK1600_ASM -DMD5_ASM -DOPENSSL_BN_ASM_GF2m -DOPENSSL_BN_ASM_MONT -DOPENSSL_BN_ASM_MONT5 -DOPENSSL_CPUID_OBJ -DOPENSSL_IA32_SSE2 -DOPENSSL_USE_APPLINK -DPADLOCK_ASM -DPOLY1305_ASM -DRC4_ASM -DSHA1_ASM -DSHA256_ASM -DSHA512_ASM -DVPAES_ASM -DWHIRLPOOL_ASM -DX25519_ASM -c /Focrypto\libcrypto-shlib-cversion.obj "crypto\cversion.c"
cversion.c
crypto\cversion.c(60): warning C4129: 'V': unrecognized character escape sequence
crypto\cversion.c(60): warning C4129: 'T': unrecognized character escape sequence
crypto\cversion.c(60): warning C4129: 'M': unrecognized character escape sequence
crypto\cversion.c(60): warning C4129: 'H': unrecognized character escape sequence
crypto\cversion.c(60): warning C4129: 'c': unrecognized character escape sequence
crypto\cversion.c(60): error C2143: syntax error: missing ';' before 'constant'
crypto\cversion.c(60): error C2143: syntax error: missing ';' before 'string'
```
</details>

I have the same issue for OpenSSL on Windows 10:

Package: openssl:x64-windows@3.1.4#1

**Host Environment**

- Host: x64-windows
- Compiler: MSVC 19.38.33133.0
-    vcpkg-tool version: 2023-12-12-1c9ec1978a6b0c2b39c9e9554a96e3e275f7556e
    vcpkg-scripts version: fatal: not a git repository: 'C:\Users\koppor\scoop\apps\vcpkg\current\.git'


**To Reproduce**

`vcpkg install openssl-windows:x64-windows`
**Failure logs**

```
-- Using cached openssl-openssl-openssl-3.1.4.tar.gz.
-- Cleaning sources at C:/Users/koppor/scoop/apps/vcpkg/current/buildtrees/openssl/src/nssl-3.1.4-c5488d00d9.clean. Use --editable to skip cleaning for the packages you specify.
-- Extracting source C:/Users/koppor/scoop/apps/vcpkg/current/downloads/openssl-openssl-openssl-3.1.4.tar.gz
-- Applying patch disable-apps.patch
-- Applying patch disable-install-docs.patch
-- Applying patch script-prefix.patch
-- Applying patch windows/install-layout.patch
-- Applying patch windows/install-pdbs.patch
-- Applying patch unix/android-cc.patch
-- Applying patch unix/move-openssldir.patch
-- Applying patch unix/no-empty-dirs.patch
-- Applying patch unix/no-static-libs-for-shared.patch
-- Using source at C:/Users/koppor/scoop/apps/vcpkg/current/buildtrees/openssl/src/nssl-3.1.4-c5488d00d9.clean
-- Found external ninja('1.11.0').
-- Getting CMake variables for x64-windows
-- Getting CMake variables for x64-windows
-- Prerunning x64-windows-dbg
CMake Error at scripts/cmake/vcpkg_execute_required_process.cmake:112 (message):
    Command failed: C:/Users/koppor/scoop/apps/vcpkg/current/downloads/tools/perl/5.38.0.1/perl/bin/perl.exe Configure enable-static-engine enable-capieng no-ssl3 no-weak-ssl-ciphers no-tests shared no-apps no-makedepend VC-WIN64A --debug --prefix=C:\\Users\\koppor\\scoop\\apps\\vcpkg\\current\\packages\\openssl_x64-windows\\debug --openssldir=C:\\Users\\koppor\\scoop\\apps\\vcpkg\\current\\packages\\openssl_x64-windows\\debug AS=C:\\Users\\koppor\\scoop\\apps\\vcpkg\\current\\downloads\\tools\\nasm\\nasm-2.16.01\\nasm.exe "CC=C:\\Program Files\\Microsoft Visual Studio\\2022\\Community\\VC\\Tools\\MSVC\\14.38.33130\\bin\\Hostx64\\x64\\cl.exe" "CFLAGS=  -nologo -DWIN32 -D_WINDOWS -W3 -utf-8 -MP  -D_DEBUG -MDd -Z7 -Ob0 -Od -RTC1  " "AR=C:\\Program Files\\Microsoft Visual Studio\\2022\\Community\\VC\\Tools\\MSVC\\14.38.33130\\bin\\Hostx64\\x64\\lib.exe" "ARFLAGS=-machine:x64 -nologo" "LD=C:\\Program Files\\Microsoft Visual Studio\\2022\\Community\\VC\\Tools\\MSVC\\14.38.33130\\bin\\Hostx64\\x64\\link.exe" "LDFLAGS= -machine:x64 -nologo    -debug -INCREMENTAL "
    Working Directory: C:/Users/koppor/scoop/apps/vcpkg/current/buildtrees/openssl/x64-windows-dbg
    Error code: 255
    See logs for more information:
      C:\Users\koppor\scoop\apps\vcpkg\current\buildtrees\openssl\prerun-specific-x64-windows-dbg-err.log

Call Stack (most recent call first):
  scripts/cmake/vcpkg_build_nmake.cmake:125 (vcpkg_execute_required_process)
  ports/openssl/windows/portfile.cmake:72 (vcpkg_build_nmake)
  ports/openssl/portfile.cmake:68 (include)
  scripts/ports.cmake:170 (include)

```
<details><summary>C:\Users\koppor\scoop\apps\vcpkg\current\buildtrees\openssl\prerun-specific-x64-windows-dbg-err.log</summary>

```

Failure!  build file wasn't produced.
Please read INSTALL.md and associated NOTES-* files.  You may also have to
look over your available compiler tool chain or change your configuration.

***** Unsupported options: no-apps
```
</details>

---

I did:

1. `scoop install vcpkg`
2. `vcpkg install openssl-windows:x64-windows`

--- a/ports/openssl/vcpkg.json
+++ b/ports/openssl/vcpkg.json
@@ -1,7 +1,7 @@
 {
   "name": "openssl",
   "version": "3.2.1",
-  "port-version": 1,
+  "port-version": 2,
   "description": "OpenSSL is an open source project that provides a robust, commercial-grade, and full-featured toolkit for the Transport Layer Security (TLS) and Secure Sockets Layer (SSL) protocols. It is also a general-purpose cryptography library.",
   "homepage": "https://www.openssl.org",
   "license": "Apache-2.0",

--- a/ports/openssl/windows/portfile.cmake
+++ b/ports/openssl/windows/portfile.cmake
@@ -1,3 +1,8 @@
+# Need cmd to pass quoted CC from nmake to mkbuildinf.pl, GH-37134
+find_program(CMD_EXECUTABLE cmd HINTS ENV PATH NO_DEFAULT_PATH REQUIRED)
+cmake_path(NATIVE_PATH CMD_EXECUTABLE cmd)
+set(ENV{COMSPEC} "${cmd}")
+
 vcpkg_find_acquire_program(PERL)
 get_filename_component(PERL_EXE_PATH "${PERL}" DIRECTORY)
 vcpkg_add_to_path("${PERL_EXE_PATH}")

--- a/versions/baseline.json
+++ b/versions/baseline.json
@@ -6482,7 +6482,7 @@
     },
     "openssl": {
       "baseline": "3.2.1",
-      "port-version": 1
+      "port-version": 2
     },
     "openssl-unix": {
       "baseline": "deprecated",

--- a/versions/o-/openssl.json
+++ b/versions/o-/openssl.json
@@ -1,6 +1,11 @@
 {
   "versions": [
     {
+      "git-tree": "4d41de10fe318363c031b97457b7b6d473020a18",
+      "version": "3.2.1",
+      "port-version": 2
+    },
+    {
       "git-tree": "3dc6df970a3aff9c977032299e481fd9872e0970",
       "version": "3.2.1",
       "port-version": 1

